### PR TITLE
dbt Templater Plugin: dbt 1.8 support

### LINKED ISSUE #5877: Add support for dbt 1.8.0
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Upgraded my virtualenv to `sqlfluff 3.0.6` and `dbt 1.8.0`. Now running `sqlfluff lint` results in an error.

### Expected Behaviour

SQLFluff to run.

### Observed Behaviour

Output of `sqlfluff lint`:

```shell
TypeError: register_adapter() missing 1 required positional argument: 'mp_context'  
```

### How to reproduce

Upgrade to mentioned package versions and run `sqlfluff lint`

### Dialect

Postgres

### Version

3.0.6

### Configuration

```
[sqlfluff]
# Supported dialects https://docs.sqlfluff.com/en/stable/dialects.html
# Or run 'sqlfluff dialects'
dialect = postgres
# One of [raw|jinja|python|placeholder]
templater = dbt
# Comma separated list of rules to exclude, or None
exclude_rules = L031
# Below controls SQLFluff output, see max_line_length for SQL output
output_line_length = 160
max_line_length = 160
# Very large files can make the parser effectively hang.
# The more efficient check is the _byte_ limit check which
# is enabled by default. The previous _character_ limit check
# is still present for backward compatibility. This will be
# removed in a future version.
# Set either to 0 to disable.
large_file_skip_byte_limit = 50000

[sqlfluff:templater:dbt]
project_dir = .
profiles_dir = ~/.dbt

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:rules]
single_table_references = unqualified

[sqlfluff:rules:capitalisation.keywords]
# Keywords
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.identifiers]
# Unquoted identifiers
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.functions]
# Function names
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.literals]
# Null & Boolean Literals
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.types]
# Data Types
extended_capitalisation_policy = lower

[sqlfluff:rules:layout.long_lines]
# Line length
ignore_comment_lines = True
ignore_comment_clauses = True

[sqlfluff:rules:references.keywords]
# Comma separated list of words to ignore for this rule
ignore_words = date,year_month,name,attribute

[sqlfluff:rules:structure.subquery]
# By default, allow subqueries in from clauses, but not join clauses
forbid_subquery_in = both

[sqlfluff:rules:ambiguous.join]
fully_qualify_join_types = both

[sqlfluff:rules:ambiguous.column_references]
# GROUP BY/ORDER BY column references
group_by_and_order_by_style = explicit

[sqlfluff:rules:references.special_chars]
# Special characters in identifiers
additional_allowed_characters = ""

[sqlfluff:rules:convention.casting_style]
# SQL type casting
preferred_type_casting_style = cast

[sqlfluff:layout:type:select_clause]
line_position = strict

[sqlfluff:layout:type:where_clause]
line_position = strict

[sqlfluff:layout:type:from_clause]
line_position = strict

[sqlfluff:layout:type:join_clause]
line_position = strict

[sqlfluff:layout:type:groupby_clause]
line_position = strict

[sqlfluff:layout:type:having_clause]
line_position = strict
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by keraion**
### Brief summary of the change made

This adds support for dbt 1.8, building off of @kzosabe's PR #5882.
- Updated plugin Dockerfile to support installation of dbt-postgres 1.8
- Added adapter protocol support calls for the decoupling of dbt-core and dbt-adapters
- Redesigned a number of test fixtures to allow for more dynamic handling
- fixes #5877

### Are there any other side effects of this change that we should be aware of?
Due to the nature of changes with dbt 1.8, this is a best effort for this to get the templater working. I may have missed things not currently covered by the tests.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py [hunk 1]
```diff
@@ -147,6 +149,11 @@ def dbt_config(self):
         from dbt.adapters.factory import register_adapter
         from dbt.config.runtime import RuntimeConfig as DbtRuntimeConfig
 
+        if self.dbt_version_tuple >= (1, 8):
+            from dbt_common.context import set_invocation_context
+
+            set_invocation_context({})
```
[the remark this answers is not in the document]
Comment (keraion) [reply plugins/sqlfluff-templater-dbt/sqlfluff_templater_dbt/templater.py:155]: This was one area that I was super uncertain about. I imagine the following should also fix the code and might be a more future-safe option.

```python
        if self.dbt_version_tuple >= (1, 8):
            from dbt_common.context import set_invocation_context
            from dbt_common.clients.system import get_env

            set_invocation_context(get_env())
```

**Review by alanmcruickshank — APPROVED** [review]
LOOKS EXCELLENT 🚀 🚀 🚀 🚀 🚀 

Thanks for much for putting this together and also tidying up the way we handle test cases in the process. 🌟

**Comment by d1manson** [comment]
thanks all for getting this sorted so quickly.

any chance of a quick release to make this available on pypi in the next day or so?... asking for a friend 😉

**Comment by kzosabe** [comment]
I read the code, it was a much more complex problem than I had imagined... Excellent work, thank you @keraion !

**Comment by elyobo** [comment]
https://github.com/sqlfluff/sqlfluff/issues/4841 is still occurring for me w/dbt 1.8.1, dbt-bigquery 1.8.1 and sqlfluff 3.0.7 which includes this fix; fine w/dbt 1.7.13, dbt-bigquery 1.7.7, and sqlfluff 3.0.7.

Should I open a new issue with this? A few 👍s on my comment on that old issue suggest it's not just me it's hitting.

**Pull request merged**